test/integration: goroutine leak check

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The current approach with ignoring a certain number of additional goroutines is too vague. During testing, we want to ensure that no additional goroutines at all get left behind by a test.

Some goroutines being started during init are okay and go-uber/goleak can handle that, in contrast to the old in-tree code. For that reason and because goleak has nicer output, it replaces the in-tree code.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
test: use go-uber/goleak for strict leak checking

It provides more readable output and has additional APIs for using it inside a
unit test. goleak.IgnoreCurrent is needed to filter out the goroutine that gets
started when importing go.opencensus.io/stats/view.

In order to handle background goroutines that get created on demand and cannot
be stopped (like the one for LogzHealth), a helper function ensures that those
are running before calling goleak.IgnoreCurrent. Keeping those goroutines
running is not a problem and thus not worth the effort of adding new APIs to
stop them.

Other goroutines are genuine leaks for which no fix is available. Those get
suppressed via IgnoreTopFunction, which works as long as that function
is unique enough.

Example output for the leak fixed in https://github.com/kubernetes/kubernetes/pull/115423:

    E0202 09:30:51.641841   74789 etcd.go:205] "EtcdMain goroutine check" err=<
        found unexpected goroutines:
        [Goroutine 4889 in state chan receive, with k8s.io/apimachinery/pkg/watch.(*Broadcaster).loop on top of the stack:
        goroutine 4889 [chan receive]:
        k8s.io/apimachinery/pkg/watch.(*Broadcaster).loop(0xc0076183c0)
        	/nvme/gopath/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch/mux.go:268 +0x65
        created by k8s.io/apimachinery/pkg/watch.NewBroadcaster
        	/nvme/gopath/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/watch/mux.go:77 +0x116
    >

pohly commented on test/integration/framework/test_server.go:
```diff
@@ -228,3 +233,10 @@ func StartTestServer(t testing.TB, setup TestServerSetup) (client.Interface, *re
 	return kubeAPIServerClient, kubeAPIServerClientConfig, tearDownFn
 }
+
+type checkNothing struct{}
```
The alternative is (untested):
```
func init() {
   _ := healtz.LogHealthz.Check(nil)
}
```

That starts the goroutine during init and then `goleak.IgnoreCurrent` knows about it.

I'm undecided whether that is a better solution than replacing the instance.

aojea replied:
can't we plumb a context on the handler to stop it on apiserver shutdown?

aojea replied:
it seems doesn't work

```
	found unexpected goroutines:
	[Goroutine 2263 in state select, with k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil on top of the stack:
	goroutine 2263 [select]:
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x59f07c8?, {0x59be860, 0xc0092efbf0}, 0x1, 0xc0000925a0)
		/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:169 +0x135
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x59c0fa0?, 0xdf8475800, 0x0, 0x25?, 0xc00317ef00?)
		/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:135 +0x89
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
		/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:92
	k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Forever(0x0?, 0x0?)
		/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:83 +0x28
	created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz.(*log).Check.func1
		/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/healthz/healthz.go:74 +0xe5
	]
```

pohly replied:
> can't we plumb a context on the handler to stop it on apiserver shutdown?

No, healthz.go initializes the variable during init. Goroutine startup then gets delayed until used for the first time.

> it seems doesn't work

I'll check that tomorrow. So far, I was testing only with `test/integration/scheduler_perf`. `make test-integration` fails for me for various reasons.

pohly replied:
We have multiple copies of `StartTestServer`. I modified the one in `test/integration/framework/test_server.go`, but not the one in `cmd/kube-apiserver/app/testing/`... let's do this all in etcd.go where the leak checking is done.

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -182,37 +182,22 @@ func EtcdMain(tests func() int) {
-		after := runtime.NumGoroutine()
-		stacktraces := make([]byte, 1<<20)
-		runtime.Stack(stacktraces, true)
-		klog.Fatalf("unexpected number of goroutines: before: %d after %d\n%sd", before, after, string(stacktraces))
```
Note that replacing this code also fixes a slight bug: the string that gets written here is `1<<20` bytes long.

The code should have used:
```
len := runtime.Stack(stacktraces, true)
... string(stacktraces[:len]) ...
```

aojea replied:
you are awesome 👏

alculquicondor commented:
this is more on sig-testing side of things
/assign @aojea

pohly commented:
> this is more on sig-testing side of things

Yes, and the change to `scheduler_perf_test.go` was by accident (too much experientation and too many rebases...). I've taken that out.

aojea commented on go.mod:
```diff
@@ -74,6 +74,7 @@ require (
 	go.opentelemetry.io/otel/sdk v1.10.0
 	go.opentelemetry.io/otel/trace v1.10.0
 	go.opentelemetry.io/proto/otlp v0.19.0
+	go.uber.org/goleak v1.2.0
```
I see this is much better but current method is simple and "good enough" for our purposes, do we want to pull a dependency just for this?  I defer to deps people to judge this

@liggitt @dims

liggitt replied:
I wouldn't want to add a dependency for this, especially not one that has any transitive dependencies

pohly replied:
> current method is simple and "good enough" for our purposes, 

It's not :disappointed: 

The current code doesn't support ignoring the goroutines that were started during init. That's crucial for reliably ignoring some of the goroutines that currently force us to the fuzzy "3 goroutine leaks are okay" thing.

> We do have some code we can reorganize into a package if needed:

I had started improving that code, but then realized that I would end up re-implementing goleak.

Is the problem with goleak the dependencies of the package? The actual code that we pull is pretty simple.

What about simply copying that code into `test/utils/goleak`?

pohly replied:
`check-dependency-stats` doesn't look too bad to me:

```
Switched to a new branch 'base'
--- /logs/artifacts/stats-base.txt	2023-02-01 16:00:51.332458491 +0000
+++ /logs/artifacts/stats.txt	2023-02-01 16:00:50.944422049 +0000
@@ -1,6 +1,6 @@
 Direct Dependencies: 210 
-Transitive Dependencies: 235 
-Total Dependencies: 285 
+Transitive Dependencies: 236 
+Total Dependencies: 286 
 Max Depth Of Dependencies: 23 
 All dependencies:
 
@@ -249,6 +249,7 @@
 go.uber.org/multierr
 go.uber.org/zap
 golang.org/x/crypto
+golang.org/x/lint
 golang.org/x/mod
 golang.org/x/net
 golang.org/x/oauth2
```

aojea replied:
we have https://github.com/kubernetes/kubernetes/tree/master/third_party/forked

pohly replied:
The integration tests are now passing with the vague "allow 3 additional goroutines" replaced with more precise suppressions through `goleak.Ignore*`. I found that in two cases I needed `goleak.IgnoreTopFunction`. I'd prefer just using goleak as-is instead of investing further effort into improving our own code or forking goleak (regardless whether that's under "third_party/forked" or "test").

@liggitt: WDYT, is `golang.org/x/lint` as new additional dependency besides goleak itself acceptable?

Or is this about other dependencies that don't show up in `go.mod` and `check-dependency-stats`?

liggitt replied:
I opened https://github.com/uber-go/goleak/issues/92 with a request / suggestion for dropping the golang.org/x/lint reference from the graph

pohly replied:
I submitted a PR for goleak. Let's wait for that.

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -182,37 +182,58 @@ func EtcdMain(tests func() int) {
+		// There's no good way to ignore this particular loop.
+		// Matching against Sleep is only indirect and not very
+		// precise.
+		goleak.IgnoreTopFunction("time.Sleep"),
```
I'm not happy with this particular Ignore. However, adding some kind of forced shutdown to https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L215-L231 also doesn't seem very appealing.

I'm open for suggestions...

aojea replied:
add a context to 
```
 recordToSink(ctx context.Context, sink EventSink, event *v1.Event, eventCorrelator *EventCorrelator, sleepDuration time.Duration)
```

and cancel the context on Shutdown

https://github.com/kubernetes/kubernetes/blob/ae23b0c11bc861bdca346666a739560ea3ee3145/staging/src/k8s.io/client-go/tools/record/event.go#L198-L200

pohly replied:
I am not sure whether that is possible. There are comments elsewhere (don't remember exactly where) that say that pending events will still get delivered after a shutdown, so the current behavior might be intentional. Changing it might break something?

pohly replied:
Besides, adding a context to `recordToSink` then implies changing the `StartEventWatcher` API - let's not.

The problem is that the sink blocks when it should simply give up:
https://github.com/kubernetes/kubernetes/blob/49ddc510bc6b57fa430f779372b71144d328995c/pkg/controller/replicaset/replica_set.go#L195

There's a TODO from 2016 about `EventSinkImpl`:

https://github.com/kubernetes/kubernetes/blame/49ddc510bc6b57fa430f779372b71144d328995c/staging/src/k8s.io/client-go/kubernetes/typed/core/v1/event_expansion.go#L154

aojea replied:
> Besides, adding a context to `recordToSink` then implies changing the `StartEventWatcher` API - let's not.

why, the context is internal and closed by `Shutdown()`, I didn't mean to expose it.

> that say that pending events will still get delivered after a shutdown

Events are ephemeral, but we can give some grace period and try to drain, WDYT?

pohly replied:
`eventBroadcasterImpl` doesn't have a context that it could pass to `recordToSink`. I guess it would be possible to create one just for this purpose. Let me try this and then we can ask others for their opinion whether it is a breaking change.

pohly replied:
Pushed an update. What I meant with "change Watch API" is that the result of `StartRecordingToSink` (a `watch.Interface`) cannot be used to shutdown the watch and the delivery callback. That is harder to change.

pohly replied:
This seems to have worked. I've turned the changed into a separate commit here: https://github.com/kubernetes/kubernetes/pull/115514

pohly commented:
/retest

aojea commented on test/integration/framework/etcd.go:
```diff
@@ -182,37 +182,58 @@ func EtcdMain(tests func() int) {
+	timeout := 5 * time.Second
+	start := time.Now()
+	for {
+		if err := goleak.Find(goleakOpts...); err != nil {
+			if time.Now().Sub(start) >= timeout {
+				klog.ErrorS(err, "EtcdMain goroutine check")
+				result = 1
+				break
+			}
+		} else {
+			break
```
one suggestion to reduce the indentation

```suggestion
		if err := goleak.Find(goleakOpts...); err == nil {
		        break
                 }
		if time.Now().Sub(start) >= timeout {
			klog.ErrorS(err, "EtcdMain goroutine check")
			result = 1
			break
		}
```

pohly replied:
That doesn't work as written, `err` won't be set after the `if`. Perhaps like this?

```
err := goleak.Find(goleakOpts...)
if err == nil {
	break
}
if time.Now().Sub(start) >= timeout {
	klog.ErrorS(err, "EtcdMain goroutine check")
	result = 1
	break
}
```

aojea replied:
oh, yeah, my bad, didn't see the error, in the klog, but you get what I mean

aojea commented:
LGTM for the replacement of the system to detect goroutine leaks, I can't say if the deps are acceptable or not, defer to the experts on the area

pohly commented on test/integration/framework/etcd.go:
```diff
@@ -182,37 +182,47 @@ func EtcdMain(tests func() int) {
+	timeout := 5 * time.Second
+	start := time.Now()
+	for {
+		err := goleak.Find(goleakOpts...)
```
This loop can get removed if https://github.com/uber-go/goleak/pull/94 gets merged.

kerthcet replied:
exactly 👍

liggitt replied:
looks like that PR got merged, can we drop this loop now?

liggitt replied:
oh, looks like that didn't make it into the 1.2.1 tag

pohly replied:
I'll keep an eye on that and update once it's in a release.

cici37 commented:
/remove-sig api-machinery

pohly commented:
I rebased to get rid of the already merged [EventBroadcaster.Shutdown](https://github.com/kubernetes/kubernetes/pull/115456/commits/5976bf38932e381c4e6a41ab30ee2afdf65ec880) commit.

I also updated to goleak master because that removes the problematic go/lint dependency. I've [asked for a proper release](https://github.com/uber-go/goleak/issues/99). Do we need to wait for that?

I'd prefer to get this PR merged soonish because it blocks further work on `scheduler_perf`.

pohly commented:
goleak v1.2.1 was release and this PR now uses it.

@liggitt: is this okay now as dependency?

If this ever causes problems during a Go update, it'll be trivial to comment out and restore later. It's only useful in tests and thus there will be no production code which could possibly depend on it.

pohly commented:
/retest

cici37 commented:
/triage accepted
